Library/Homebrew/command.rb

```diff
@@ -0,0 +1,212 @@
+      invalid_option_string = "#{invalid_option_pluralize} provided: #{argv_invalid_options.join " "}"
+      unless argv_options_without_value.empty?
+        invalid_option_string = <<-EOS.undent
+          #{invalid_option_string}
+          #{argv_options_without_value.map { |k, v| "#{k} requires a value <#{v}>" }.join("\n")}
+        EOS
+      end
+      error_message = nil
+      if @valid_options.empty?
+        error_message = <<-EOS.undent
+          #{invalid_option_string}
+          The command has no valid options
+
+        EOS
+      else
+        error_message = <<-EOS.undent
+          #{invalid_option_string}
+          Correct usage:
+          #{@help_output}
+        EOS
+      end
+      error_message
+    end
+
+    def self.check_invalid_options
+      error_message = get_error_message(ARGV.options_only)
+      odie error_message unless error_message.nil?
+    end
+
+    def self.option_string(option)
+      hash = @valid_options.find { |x| x[:option] == option }
+      child_options = hash[:child_options]
+
+      # return "[`#{option}`]" if child_options.nil?
+      if child_options.nil?
+        value = hash[:value]
+        return "[`#{option}`]" if value.nil?
+        return "[`#{option}=`<#{value}>]"
+      end
+
+      childs_str = ""
+      child_options.each do |child_option|
+        child_str = option_string(child_option)
+        childs_str += child_str
+      end
+      "[`#{option}` #{childs_str}]"
+    end
+
+    def self.desc_string(option, begin_spaces = 4, parent_present = false)
+      hash = @valid_options.find { |x| x[:option] == option }
+      desc = hash[:desc]
+      child_options = hash[:child_options]
+      if child_options.nil?
+        # return " "*begin_spaces + "With `#{option}`, #{desc}\n" if parent_present
+        # return " "*begin_spaces + "If `#{option}` is passed, #{desc}\n"
+        option_value = hash[:value]
+        if parent_present
+          return " "*begin_spaces + "With `#{option}`, #{desc}\n" if option_value.nil?
+          return " "*begin_spaces + "With `#{option}=`<#{option_value}>, #{desc}\n"
+        else
+          return " "*begin_spaces + "If `#{option}` is passed, #{desc}\n" if option_value.nil?
+          return " "*begin_spaces + "If `#{option}=`<#{option_value}> is specified, #{desc}\n"
+        end
+      else
+        childs_str = ""
+        child_options.each do |child_option|
+          # TODO: change begin_spaces to begin_spaces+2 if maintainers agree on indenting the descriptions of childs
+          child_str = desc_string(child_option, begin_spaces, true)
+          childs_str += child_str
+        end
+        return " "*begin_spaces + "With `#{option}`, #{desc}\n#{childs_str}" if parent_present
+        return " "*begin_spaces + "If `#{option}` is passed, #{desc}\n#{childs_str}"
+      end
+    end
+
+    def self.generate_help_and_manpage_output
+      option_str = ""
+      desc_str = ""
+      @root_options.each do |root_option|
```

Comment:
Same thing here, use one `map` and `join` combination for `desc_str` and one for `option_str`.